### ISSUE TITLE
Private Method

### ISSUE THREAD

**Issue by cgarciae** [billysometimes/rethinkdb#14]
How can you run `c._start(...)` in
```dart
  Future run(Connection c, [global_optargs]){
      if(c == null)
        throw new RqlDriverError("RqlQuery.run must be given a connection to run on.");
      if(global_optargs == null)
        global_optargs = {};
      return c._start(this, global_optargs);
    }
```
if its supposed to be a private method?

**Comment by billysometimes** [comment]
Yes, it should be private.  it submits the query on the connection.  The user should do that by calling the query's run method, just like in the javascript driver:
```dart
 r.dbList().run(connection).then((List response){
    //do stuff here
});
```

_start will be called by the run method.

**Comment by cgarciae** [comment]
@billysometimes Yeah, but why doesn't it throw any erros? Supposedly you cant access field/methods that begin with `_`

**Comment by billysometimes** [comment]
Private variables in dart are restricted by library, not by class.  Since Connection and Query are both part of the same library, they can access each other's methods, even if they start with an underscore.

which is very convenient in cases like this where we want our library to use that method without giving the user the ability to call that method.

**Comment by cgarciae** [comment]
@billysometimes Thanks for the info. I am creating a Rethink Plugin for Redstone:
https://github.com/redstone-dart/redstone_rethinkdb

**Comment by cgarciae** [comment]
@billysometimes I am having problems with this, if you use methods with underscore it become impossible for others to implement your interfaces. Private methods are not part of an interface but they are required by your library.

**Comment by billysometimes** [comment]
I'm afraid I don't follow, the only methods that should be public from Connection are:

connect
close
reconnect
use
noreplyWait
on
addListener

What is the problem you are having? I can try to help out.

**Comment by cgarciae** [comment]
@billysometimes 
I have this class
```dart
class RethinkConnection extends ManagedConnection<Connection> implements Connection {

  RethinkConnection(int connId, Connection conn) : super (connId, conn);

  get isClosed {
    return conn.isClosed;
  }

  noreplyWait() {
    conn.noreplyWait();
  }
...//you get the idea
}
```
It extends `ManagedConnection<Connection>` to play nice with both (your own) `Rbd` and (a class from Redstone called) `DatabaseManager<ManagedConnection<Connection>>`. Both happily depend on `ConnectionPool` so everything works like a dream.
Because of how Redstone's mappers plugin manages connections, you get a `ManagedConnection<Connection>` (RethinkConnection implements that type) injected into your route
```dart
@mvc.DataController ('/rethink', methods: const[app.POST], allowMultipartRequest: true)
rethink (@app.Attr("dbConn") RethinkConnection conn,  @Decode(from: const[app.FORM]) User user) async {

  var resp = await r.table("usuarios")
    .insert(encode(user))
    .run(conn);
```

I didn't want me or the user of the library to have to type ``.run(conn.conn)` because its both too wordy and not standard. So I made `RethinkConnection` a light proxy of `Connection`. It technically works, except that the `RqlQuery.run` method calls the private `Connection._start` method which my library does not implement. So there are 2 options:

1. I implement the private method (with the worry that maybe a I'll end implementing the whole class later.
2. You either make `start` a public method or access private members from other objects.

**Comment by billysometimes** [comment]
I see.

I don't think option 2 will happen, `start` shouldn't be public.

Some other options to consider might be to override redstones `getMapperPlugin ` so that it sets 
`app.request.attributes["dbConn"] = conn.conn;`

or add another interceptor that changes the request attribute if it exists:
```dart
if(app.request.attributes.containsKey("dbConn")){
  app.request.attributes["dbConn"] = app.request.attributes["dbConn"].conn;
}
```
or follow the pattern of the [postgres mapper](https://github.com/redstone-dart/redstone_mapper_pg)
```dart 
Connection get conn => app.request.attributes.dbConn.conn;
```

**Comment by cgarciae** [comment]
Ok. I sort of did something similar, I'll share what I implemented. Its looking very nice.

**Comment by billysometimes** [comment]
sounds great! I look forward to playing around with it.
